### PR TITLE
Fix bug where finalizeRekey wasn't being called

### PR DESCRIPTION
**Pull request by akalin-keybase**
Move TLFCryptKey encryption logic to
AddKeyGeneration. Also clean up those interfaces.

Rename AddKeysForTest to addKeyGenerationForTest.

Clarified interface comments for FinalizeRekey.

Remove key cache mock from tests.

Fix a few test hangs on errors.

### CODE REVIEW

File: libkbfs/key_manager.go [hunk 1]
```diff
@@ -733,24 +757,32 @@ func (km *KeyManagerStandard) Rekey(ctx context.Context, md *RootMetadata, promp
...
+
 	if !isWriter {
 		if len(newReaderUsers) > 0 || addNewWriterDevice || incKeyGen {
 			// If we're a reader but we haven't completed all the work, return
 			// RekeyIncompleteError.
 			return addNewReaderDeviceForSelf, nil, RekeyIncompleteError{}
 		}
 		// Otherwise, there's nothing left to do!
-		if err := md.finalizeRekey(km.config.Crypto(),
-			kbfscrypto.TLFCryptKey{}, tlfCryptKey); err != nil {
```
Comment (akalin-keybase) [inline libkbfs/key_manager.go]: these shouldn't be passing in empty keys for prevTLFCryptKey, right?

**Comment by akalin-keybase** [comment]
PTAL, both of you!

**Comment by akalin-keybase** [comment]
Also, forgot to mention that some of the changes are covered by existing tests here, and some will be covered by a future PR (which tests everything with both MD versions).

**Comment by aalness** [comment]
LGTM! Thanks Fred.

File: libkbfs/key_manager.go [hunk 2]
```diff
@@ -733,24 +757,32 @@ func (km *KeyManagerStandard) Rekey(ctx context.Context, md *RootMetadata, promp
 		}
 	}()
 
+	// From this point on, if we return true for mdChanged with a
+	// nil error or RekeyIncompleteError{}, we must call
+	// km.finalizeRekey(md) first.
+
+	defer func() {
+		if mdChanged && (err == nil || err == RekeyIncompleteError{}) {
```
Comment (jzila) [inline libkbfs/key_manager.go]: this `RekeyIncompleteError` check could be brittle if we ever add fields to it; maybe check the type instead?

File: libkbfs/key_manager.go (hunk 1 above)
Comment (akalin-keybase) [inline libkbfs/key_manager.go]: Ah, right. I'll rework this PR, then...

Comment (akalin-keybase) [inline libkbfs/key_manager.go]: Okay, reworked this by moving all the TLFCryptKey encryption logic into AddKeyGeneration, which is where it should be anyway.

**Comment by akalin-keybase** [comment]
@jzila PTAL, since I reworked this PR substantially. Thanks!

File: libkbfs/key_manager.go (hunk 2 above)
Comment (akalin-keybase) [inline libkbfs/key_manager.go]: Oops, forgot to address this. Fixed now.

**Comment by aalness** [comment]
👍

**Pull request merged**